src/init.c

```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+	if (array == NULL) {
+		fprintf(stderr, "can not convert line to array: %s\n", new_line);
+		goto out;
+	}
+
+	for (size_t i=0; array[i] != NULL; i++) {
+		if (strncmp(array[i], pid_str, pid_str_size) == 0) {
+			ret = i;
+			break;
+		}
+	}
+
+out:
+	free_if_set(new_line);
+	free_if_set(array);
+	return ret;
+}
+
+/**
+ * appends current ps line to final ps output
+ *
+ * \param[in,out] ps_data \ref hyper_ps_data
+ *
+ * \return true on success, else false
+ */
+static bool hyper_append_ps_line(struct hyper_ps_data *ps_data)
+{
+	/* +1 for \n */
+	size_t line_len = ps_data->line_len + 1;
+	size_t output_len = *ps_data->output_len;
+	const size_t buf_size = 1024;
+	int ret = -1;
+
+	if (ps_data->output_cap < line_len) {
+		uint8_t *tmp = NULL;
+
+		tmp = (uint8_t *)realloc(*ps_data->output,
+			(ps_data->output_cap + output_len + buf_size) * sizeof(**ps_data->output));
+
+		if (tmp == NULL) {
+			/* try to reallocate once again */
+			tmp = (uint8_t *)realloc(*ps_data->output,
+				(ps_data->output_cap + output_len + buf_size) * sizeof(**ps_data->output));
+		}
+
+		if (tmp == NULL) {
+			return false;
+		}
+
+		*ps_data->output = tmp;
+		ps_data->output_cap += buf_size;
+	}
+
+	ret = snprintf((char*)(*ps_data->output + output_len),
+		line_len + 1, "%s\n", ps_data->line);
+	if (ret < 0) {
+		return false;
+	}
+
+	ps_data->output_cap -= ret;
+	*ps_data->output_len += ret;
+
+	return true;
+}
+
+/**
+ * process each ps line:
+ * - if pid is equal to container pid then add current line to the
+ *   final ps output (ps_data->output)
+ * - if pid is greater than container pid then it could be running inside
+ *   the container, else skip it
+ * - check if pid is running in the same container mount namespace,
+ *   if it is then add current line to the final ps output (ps_data->output)
+ *
+ * \param[in,out] ps_data \ref hyper_ps_data
+ *
+ * \return true on success, else false
+ */
+static bool hyper_process_ps_line(struct hyper_ps_data *ps_data)
+{
+	char *new_process_line = NULL;
+	char **array = NULL;
+	size_t array_size = 0;
+	pid_t pid;
+	char mnt_ns_path[PATH_MAX];
+	struct stat st;
+
+	new_process_line = strdup(ps_data->line);
+	if (new_process_line == NULL) {
+		fprintf(stderr, "can not dup string '%s': %s\n",
+			ps_data->line, strerror(errno));
+		return false;
+	}
+
+	array = hyper_line_to_array(new_process_line, &array_size, ps_data->pid_index+1);
+	if (array == NULL) {
+		free_if_set(new_process_line);
+		fprintf(stderr, "can not convert line to array: %s\n", new_process_line);
+		return false;
+	}
+
+	if (ps_data->pid_index >= array_size) {
+		/* current line has not a pid */
+		goto out;
+	}
+
+	/* get pid from line */
+	pid = strtol(array[ps_data->pid_index], NULL, 10);
+	if (pid == 0) {
+		fprintf(stderr, "can not convert string '%s' to int: %s\n",
+			array[ps_data->pid_index], strerror(errno));
+		goto out;
+	}
+
+	/* processes less than container pid are not part of the container */
+	if (pid < ps_data->container_pid) {
```

Comment:
Is this test reliable I wonder? The `container_pid` is in a separate namespace so might be `1` inside the container and `999` outside it in the VM context.

What if the container creates lots of processes and reaches the limit of a `pid_t`. Will the next pid be guaranteed to be `> container_pid` (`>999` in the VM context in this example)?